lib/shader/dspbr.glsl

```diff
@@ -13,58 +13,120 @@
+float sqr(float x)
 {
-    float ax = alpha.x * wi.x;
-    float ay = alpha.y * wi.y;
+    return x * x;
+}
+
+/////////////////////////////////////////////////
+// GGX Distribution
 
-    float inv_a_2 = (ax * ax + ay * ay) / (wi.z* wi.z);
+// Eric Heitz. Understanding the Masking-Shadowing Function in Microfacet-Based BRDFs.
+// Journal of Computer Graphics TechniquesVol. 3, No. 2, 2014
+// http://jcgt.org/published/0003/02/03/paper.pdf
 
-    return 2.0 / (1.0 + sqrt(1.0 + inv_a_2));
+// eq. 80
+float projected_roughness(vec2 alpha, vec3 w, Geometry g)
+{
+    float sin_theta_2 = 1.0 - sqr(dot(w, g.n));
+    float inv_sin_theta_2 = 1.0 / sin_theta_2;
+    float cos_phi_2 = sqr(dot(w, g.t)) * inv_sin_theta_2;
+    float sin_phi_2 = sqr(dot(w, g.b)) * inv_sin_theta_2;
+
+    return sqrt(cos_phi_2 * sqr(alpha.x) + sin_phi_2 * sqr(alpha.y));    
 }
 
-float ggx_g1(in vec2 alpha, in vec3 w, in vec3 wh)
+// eq. 86
+float ggx_smith_lambda(vec2 alpha, vec3 w, Geometry g)
 {
-    if (dot(w, wh) <= 0.0 || w.z == 0.0) {
+    float sin_theta_2 = 1.0 - sqr(dot(w, g.n));
+
+    if (sin_theta_2 < 0.0001) {
         return 0.0;
-    } else {
-        vec3 alphauv1 = vec3(alpha.x, alpha.y, 1.0);
-        float abs_cos_w = abs(w.z);
-        return 2.0 * abs_cos_w / (length(alphauv1 * w) + abs_cos_w);
     }
+
+    float alpha_w = projected_roughness(alpha, w, g);
+
+    float tan_theta = sqrt(sin_theta_2) / abs(dot(w, g.n));
+    float a = 1.0 / (alpha_w * tan_theta);
+
+    return 0.5 * (-1.0 + sqrt(1.0 + 1.0 / sqr(a)));
 }
 
-// Height-Correlated Masking and Shadowing
-// equation 21 in 'Understanding the Masking-Shadowing Function in Microfacet-Based BRDFs'
-float ggx_g(in vec2 alpha, in vec3 wi, in vec3 wo, in vec3 wh)
+// eq. 43
+float ggx_smith_g1(vec2 alpha, vec3 w, vec3 wh, Geometry g)
 {
-    bool has_no_contribution = dot(wi, wh) <= 0.0 || dot(wo, wh) <= 0.0;
-    if(has_no_contribution || (wi.z == 0.0 && wo.z == 0.0))
+    if (dot(w, wh) < 0.0) {
         return 0.0;
-    else {
-        float abs_cos_wi = abs(wi.z);
-        float abs_cos_wo = abs(wo.z);
+    }
+
+    return 1.0 / (1.0 + ggx_smith_lambda(alpha, w, g));
+}
 
-        vec3 alphauv1 = vec3(alpha.x, alpha.y, 1.0);
-        return 2.0 * abs_cos_wi * abs_cos_wo / (length(alphauv1 * wi) * abs_cos_wo + length(alphauv1 * wo) * abs_cos_wi);
+// eq. 99
+float ggx_smith_g2(vec2 alpha, vec3 wi, vec3 wo, vec3 wh, Geometry g)
+{
+    if (dot(wo, wh) * dot(wi, wh) < 0.0) {
+        return 0.0;
     }
-    // float G1 = smith_lambda(wi, alpha);
-    // float G2 = smith_lambda(wo, alpha);
-    // return G1 * G2;
+
+    float lambda_wi = ggx_smith_lambda(alpha, wi, g);
+    float lambda_wo = ggx_smith_lambda(alpha, wo, g);
+
+    return 1.0 / (1.0 + lambda_wi + lambda_wo);
 }
 
-// Heitz 2017, A Simpler and Exact Sampling Routine for the GGX Distribution of Visible Normals
+// GGX distribution, eq. 85
+float ggx_eval(vec2 alpha, vec3 wh, Geometry g)
+{
+    if (dot(wh, g.n) < 0.0) {
```

Comment:
Fixes division by zero and removes fireflys for very small angles
```suggestion
    if (dot(wh, g.n) < EPS_COS) {
```

Reply:
Thanks!